File: apps/desktop/src/ipc/lsp.ts
import { invoke } from '@tauri-apps/api/core';
import type { LspCompletionItem, LspHoverResult, LspLocation, LspSymbol } from '@forge/core';

export const lspIpc = {
  start: (bayId: string, languageId: string, rootPath: string): Promise<void> =>
    invoke('lsp_start', { bayId, languageId, rootPath }),
  stop: (bayId: string, languageId: string): Promise<void> =>
    invoke('lsp_stop', { bayId, languageId }),
  stopAll: (bayId: string): Promise<void> => invoke('lsp_stop_all', { bayId }),
  didOpen: (bayId: string, languageId: string, uri: string, text: string): Promise<void> =>
    invoke('lsp_did_open', { bayId, languageId, uri, text }),
  didChange: (
    bayId: string,
    languageId: string,
    uri: string,
    version: number,
    text: string,
  ): Promise<void> => invoke('lsp_did_change', { bayId, languageId, uri, text, version }),
  completion: (
    bayId: string,
    languageId: string,
    uri: string,
    line: number,
    character: number,
  ): Promise<LspCompletionItem[]> =>
    invoke('lsp_completion', { bayId, languageId, uri, line, character }),
  hover: (
    bayId: string,
    languageId: string,
    uri: string,
    line: number,
    character: number,
  ): Promise<LspHoverResult | null> =>
    invoke('lsp_hover', { bayId, languageId, uri, line, character }),
  definition: (
    bayId: string,
    languageId: string,
    uri: string,
    line: number,
    character: number,
  ): Promise<LspLocation[] | null> =>
    invoke('lsp_definition', { bayId, languageId, uri, line, character }),
  documentSymbols: (bayId: string, languageId: string, uri: string): Promise<LspSymbol[]> =>
    invoke('lsp_document_symbols', { bayId, languageId, uri }),
  workspaceSymbols: (bayId: string, languageId: string, query: string): Promise<LspSymbol[]> =>
    invoke('lsp_workspace_symbols', { bayId, languageId, query }),
};

File: apps/desktop/src-tauri/src/commands/lsp.rs
use tauri::State;

use crate::lsp::manager::LspManager;
use crate::lsp::types::*;

#[tauri::command]
pub async fn lsp_start(
    bay_id: String,
    language_id: String,
    root_path: String,
    app: tauri::AppHandle,
    state: State<'_, LspManager>,
) -> Result<(), String> {
    state
        .start_server(&bay_id, &language_id, &root_path, app)
        .await
}

#[tauri::command]
pub async fn lsp_stop(
    bay_id: String,
    language_id: String,
    state: State<'_, LspManager>,
) -> Result<(), String> {
    state.stop_server(&bay_id, &language_id).await
}

#[tauri::command]
pub async fn lsp_stop_all(
    bay_id: String,
    state: State<'_, LspManager>,
) -> Result<(), String> {
    state.stop_all(&bay_id).await
}

#[tauri::command]
pub async fn lsp_did_open(
    bay_id: String,
    language_id: String,
    uri: String,
    text: String,
    state: State<'_, LspManager>,
) -> Result<(), String> {
    state
        .did_open(&bay_id, &language_id, &uri, &text)
        .await
}

#[tauri::command]
pub async fn lsp_did_change(
    bay_id: String,
    language_id: String,
    uri: String,
    version: i32,
    text: String,
    state: State<'_, LspManager>,
) -> Result<(), String> {
    state
        .did_change(&bay_id, &language_id, &uri, version, &text)
        .await
}

#[tauri::command]
pub async fn lsp_completion(
    bay_id: String,
    language_id: String,
    uri: String,
    line: u32,
    character: u32,
    state: State<'_, LspManager>,
) -> Result<Vec<LspCompletionItem>, String> {
    state
        .completion(&bay_id, &language_id, &uri, line, character)
        .await
}

#[tauri::command]
pub async fn lsp_hover(
    bay_id: String,
    language_id: String,
    uri: String,
    line: u32,
    character: u32,
    state: State<'_, LspManager>,
) -> Result<Option<LspHoverResult>, String> {
    state
        .hover(&bay_id, &language_id, &uri, line, character)
        .await
}

#[tauri::command]
pub async fn lsp_definition(
    bay_id: String,
    language_id: String,
    uri: String,
    line: u32,
    character: u32,
    state: State<'_, LspManager>,
) -> Result<Option<Vec<LspLocation>>, String> {
    state
        .definition(&bay_id, &language_id, &uri, line, character)
        .await
}

#[tauri::command]
pub async fn lsp_document_symbols(
    bay_id: String,
    language_id: String,
    uri: String,
    state: State<'_, LspManager>,
) -> Result<Vec<LspSymbol>, String> {
    state
        .document_symbol(&bay_id, &language_id, &uri)
        .await
}

#[tauri::command]
pub async fn lsp_workspace_symbols(
    bay_id: String,
    language_id: String,
    query: String,
    state: State<'_, LspManager>,
) -> Result<Vec<LspSymbol>, String> {
    state
        .workspace_symbol(&bay_id, &language_id, &query)
        .await
}
